In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('credit_risk_dataset.csv')

In [3]:
df.head()

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length
0,22,59000,RENT,123.0,PERSONAL,D,35000,16.02,1,0.59,Y,3
1,21,9600,OWN,5.0,EDUCATION,B,1000,11.14,0,0.10,N,2
2,25,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87,1,0.57,N,3
3,23,65500,RENT,4.0,MEDICAL,C,35000,15.23,1,0.53,N,2
4,24,54400,RENT,8.0,MEDICAL,C,35000,14.27,1,0.55,Y,4


In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 32581 entries, 0 to 32580
Data columns (total 12 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   person_age                  32581 non-null  int64  
 1   person_income               32581 non-null  int64  
 2   person_home_ownership       32581 non-null  str    
 3   person_emp_length           31686 non-null  float64
 4   loan_intent                 32581 non-null  str    
 5   loan_grade                  32581 non-null  str    
 6   loan_amnt                   32581 non-null  int64  
 7   loan_int_rate               29465 non-null  float64
 8   loan_status                 32581 non-null  int64  
 9   loan_percent_income         32581 non-null  float64
 10  cb_person_default_on_file   32581 non-null  str    
 11  cb_person_cred_hist_length  32581 non-null  int64  
dtypes: float64(3), int64(5), str(4)
memory usage: 3.5 MB


In [5]:
df.isnull().sum()

person_age                       0
person_income                    0
person_home_ownership            0
person_emp_length              895
loan_intent                      0
loan_grade                       0
loan_amnt                        0
loan_int_rate                 3116
loan_status                      0
loan_percent_income              0
cb_person_default_on_file        0
cb_person_cred_hist_length       0
dtype: int64

##### person_emp_length has 895 and loan_int_rate has 3116 missing values

In [6]:
df.describe()

,person_age,person_income,person_emp_length,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_cred_hist_length
count,32581.000000,3.258100e+04,31686.000000,32581.000000,29465.000000,32581.000000,32581.000000,32581.000000
mean,27.734600,6.607485e+04,4.789686,9589.371106,11.011695,0.218164,0.170203,5.804211
std,6.348078,6.198312e+04,4.142630,6322.086646,3.240459,0.413006,0.106782,4.055001
min,20.000000,4.000000e+03,0.000000,500.000000,5.420000,0.000000,0.000000,2.000000
25%,23.000000,3.850000e+04,2.000000,5000.000000,7.900000,0.000000,0.090000,3.000000
50%,26.000000,5.500000e+04,4.000000,8000.000000,10.990000,0.000000,0.150000,4.000000
75%,30.000000,7.920000e+04,7.000000,12200.000000,13.470000,0.000000,0.230000,8.000000
max,144.000000,6.000000e+06,123.000000,35000.000000,23.220000,1.000000,0.830000,30.000000


##### person age has outliers 144 is impossible to become a age of a person.
##### person income has also outlier. 6000000 income must be outlier.
##### person employee length 123 max which is outlier

In [7]:
df['person_age'].value_counts()

person_age
23     3889
22     3633
24     3549
25     3037
26     2477
27     2138
28     1854
29     1687
30     1316
21     1229
31     1142
32      964
33      856
34      709
35      620
36      548
37      478
38      373
39      302
40      271
41      241
42      188
43      164
44      141
45      108
46       94
47       94
48       75
50       52
49       49
51       39
52       36
53       30
54       24
55       20
58       19
20       15
57       15
60       15
56       15
65        9
66        9
61        9
62        7
64        7
70        7
59        5
69        5
144       3
73        3
63        3
123       2
78        1
94        1
80        1
84        1
76        1
67        1
Name: count, dtype: int64

In [8]:
# we have to drop ages that are greater than 100 years
df[df['person_age'] > 100]

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length
81,144,250000,RENT,4.0,VENTURE,C,4800,13.57,0,0.02,N,3
183,144,200000,MORTGAGE,4.0,EDUCATION,B,6000,11.86,0,0.03,N,2
575,123,80004,RENT,2.0,EDUCATION,B,20400,10.25,0,0.25,N,3
747,123,78000,RENT,7.0,VENTURE,B,20000,NaN,0,0.26,N,4
32297,144,6000000,MORTGAGE,12.0,PERSONAL,C,5000,12.73,0,0.00,N,25


In [9]:
df = df[df['person_age'] <= 100]

In [10]:
df['person_emp_length'].value_counts()

person_emp_length
0.0      4105
2.0      3848
3.0      3456
5.0      2946
1.0      2915
4.0      2872
6.0      2666
7.0      2195
8.0      1687
9.0      1367
11.0      740
10.0      696
12.0      574
13.0      426
14.0      335
15.0      238
16.0      165
17.0      129
18.0      104
19.0       64
20.0       42
21.0       38
22.0       19
24.0       10
23.0       10
25.0        8
26.0        6
27.0        5
31.0        4
28.0        3
123.0       2
30.0        2
41.0        1
34.0        1
29.0        1
38.0        1
Name: count, dtype: int64

In [11]:
df = df[(df['person_emp_length'].isnull()) | (df['person_emp_length'] < 60)]

In [12]:
df['person_emp_length'] = df['person_emp_length'].fillna(df['person_emp_length'].median())

In [13]:
df['person_emp_length'].isnull().sum()

np.int64(0)

In [14]:
df['loan_int_rate'].describe()

count    29459.000000
mean        11.011529
std          3.240497
min          5.420000
25%          7.900000
50%         10.990000
75%         13.470000
max         23.220000
Name: loan_int_rate, dtype: float64

In [15]:
df['loan_int_rate'] = df.groupby('loan_grade')['loan_int_rate'].transform(
    lambda x: x.fillna(x.median())
)

In [16]:
df.isnull().sum()

person_age                    0
person_income                 0
person_home_ownership         0
person_emp_length             0
loan_intent                   0
loan_grade                    0
loan_amnt                     0
loan_int_rate                 0
loan_status                   0
loan_percent_income           0
cb_person_default_on_file     0
cb_person_cred_hist_length    0
dtype: int64

In [17]:
df['person_income'].quantile(0.99)

np.float64(225000.0)

In [18]:
# fresher and parttime job holders imcomes are less than other so i use only upper bounds to handle outliers
upper = df['person_income'].quantile(.99)
df['person_income'] = df['person_income'].clip(upper=upper)

In [19]:
df.describe()

,person_age,person_income,person_emp_length,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_cred_hist_length
count,32574.000000,32574.000000,32574.000000,32574.000000,32574.000000,32574.000000,32574.000000,32574.000000
mean,27.718426,64315.123780,4.760576,9588.018051,11.013752,0.218180,0.170202,5.804108
std,6.204987,37766.523247,3.981181,6320.249598,3.212328,0.413017,0.106755,4.053873
min,20.000000,4000.000000,0.000000,500.000000,5.420000,0.000000,0.000000,2.000000
25%,23.000000,38500.000000,2.000000,5000.000000,7.880000,0.000000,0.090000,3.000000
50%,26.000000,55000.000000,4.000000,8000.000000,10.990000,0.000000,0.150000,4.000000
75%,30.000000,79200.000000,7.000000,12200.000000,13.480000,0.000000,0.230000,8.000000
max,94.000000,225000.000000,41.000000,35000.000000,23.220000,1.000000,0.830000,30.000000


In [20]:
df.dtypes

person_age                      int64
person_income                   int64
person_home_ownership             str
person_emp_length             float64
loan_intent                       str
loan_grade                        str
loan_amnt                       int64
loan_int_rate                 float64
loan_status                     int64
loan_percent_income           float64
cb_person_default_on_file         str
cb_person_cred_hist_length      int64
dtype: object

In [21]:
df.head()

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length
1,21,9600,OWN,5.0,EDUCATION,B,1000,11.14,0,0.10,N,2
2,25,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87,1,0.57,N,3
3,23,65500,RENT,4.0,MEDICAL,C,35000,15.23,1,0.53,N,2
4,24,54400,RENT,8.0,MEDICAL,C,35000,14.27,1,0.55,Y,4
5,21,9900,OWN,2.0,VENTURE,A,2500,7.14,1,0.25,N,2


In [22]:
df['loan_status_label'] = df['loan_status'].replace({0: "Non-Default",
                                                1: "Default"})

In [23]:
df.tail()

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length,loan_status_label
32576,57,53000,MORTGAGE,1.0,PERSONAL,C,5800,13.16,0,0.11,N,30,Non-Default
32577,54,120000,MORTGAGE,4.0,PERSONAL,A,17625,7.49,0,0.15,N,19,Non-Default
32578,65,76000,RENT,3.0,HOMEIMPROVEMENT,B,35000,10.99,1,0.46,N,28,Default
32579,56,150000,MORTGAGE,5.0,PERSONAL,B,15000,11.48,0,0.10,N,26,Non-Default
32580,66,42000,RENT,2.0,MEDICAL,B,6475,9.99,0,0.15,N,30,Non-Default


In [24]:
labels = ['20-25','26-30','31-40','41-50','50+']
df['age_group'] = pd.cut(df['person_age'], bins= [20,25,30,40,50,100], labels=labels)

In [25]:
df['person_income'].value_counts(bins=5)

(48200.0, 92400.0]      14194
(3778.999, 48200.0]     12905
(92400.0, 136600.0]      3803
(136600.0, 180800.0]     1019
(180800.0, 225000.0]      653
Name: count, dtype: int64

In [26]:
bins=[0,48000,92000,136000,225000]
labels = ['Low','Medium','High','Very High']
df['income_group'] = pd.cut(df['person_income'], bins = bins, labels= labels)

In [27]:
df['loan_amnt'].value_counts(bins = 5)

(465.499, 7400.0]     14579
(7400.0, 14300.0]     11273
(14300.0, 21200.0]     4788
(21200.0, 28100.0]     1597
(28100.0, 35000.0]      337
Name: count, dtype: int64

In [28]:
bins = [0,7400, 14300, 21200, 35000]
labels = ['Small','Medium','Large','Very Large']
df['loan_group'] = pd.cut(df['loan_amnt'], bins=bins, labels= labels)

In [29]:
df.head()

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length,loan_status_label,age_group,income_group,loan_group
1,21,9600,OWN,5.0,EDUCATION,B,1000,11.14,0,0.10,N,2,Non-Default,20-25,Low,Small
2,25,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87,1,0.57,N,3,Default,20-25,Low,Small
3,23,65500,RENT,4.0,MEDICAL,C,35000,15.23,1,0.53,N,2,Default,20-25,Medium,Very Large
4,24,54400,RENT,8.0,MEDICAL,C,35000,14.27,1,0.55,Y,4,Default,20-25,Medium,Very Large
5,21,9900,OWN,2.0,VENTURE,A,2500,7.14,1,0.25,N,2,Default,20-25,Low,Small


In [30]:
df.rename(columns={
    'person_age':'age',
    'person_income':'income',
    'person_home_ownership':'home_ownership',
    'person_emp_length':'emp_length',
    'loan_amnt' : 'loan_amount',
    'loan_int_rate':'interest_rate',
    'loan_percent_income':'loan_to_income',
    'cb_person_default_on_file':'prev_default',
    'cb_person_cred_hist_length':'credit_hist_length'
}, inplace=True)

In [31]:
df.head()

,age,income,home_ownership,emp_length,loan_intent,loan_grade,loan_amount,interest_rate,loan_status,loan_to_income,prev_default,credit_hist_length,loan_status_label,age_group,income_group,loan_group
1,21,9600,OWN,5.0,EDUCATION,B,1000,11.14,0,0.10,N,2,Non-Default,20-25,Low,Small
2,25,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87,1,0.57,N,3,Default,20-25,Low,Small
3,23,65500,RENT,4.0,MEDICAL,C,35000,15.23,1,0.53,N,2,Default,20-25,Medium,Very Large
4,24,54400,RENT,8.0,MEDICAL,C,35000,14.27,1,0.55,Y,4,Default,20-25,Medium,Very Large
5,21,9900,OWN,2.0,VENTURE,A,2500,7.14,1,0.25,N,2,Default,20-25,Low,Small


In [32]:
df['prev_default'] = df['prev_default'].map({'Y':'Yes',
                                            'N':'No'})

In [33]:
df['loan_risk'] = df['loan_grade'].map({
    'A': 'Low Risk',
    'B': 'Low Risk', 
    'C': 'Medium Risk',
    'D': 'Medium Risk',
    'E': 'High Risk',
    'F': 'High Risk',
    'G': 'High Risk'
})

In [34]:
df['interest_rate_group'] = pd.cut(df['interest_rate'],
    bins=[0, 8, 12, 16, 25],
    labels=['Low', 'Medium', 'High', 'Very High'])

In [35]:
df.head()

,age,income,home_ownership,emp_length,loan_intent,loan_grade,loan_amount,interest_rate,loan_status,loan_to_income,prev_default,credit_hist_length,loan_status_label,age_group,income_group,loan_group,loan_risk,interest_rate_group
1,21,9600,OWN,5.0,EDUCATION,B,1000,11.14,0,0.10,No,2,Non-Default,20-25,Low,Small,Low Risk,Medium
2,25,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87,1,0.57,No,3,Default,20-25,Low,Small,Medium Risk,High
3,23,65500,RENT,4.0,MEDICAL,C,35000,15.23,1,0.53,No,2,Default,20-25,Medium,Very Large,Medium Risk,High
4,24,54400,RENT,8.0,MEDICAL,C,35000,14.27,1,0.55,Yes,4,Default,20-25,Medium,Very Large,Medium Risk,High
5,21,9900,OWN,2.0,VENTURE,A,2500,7.14,1,0.25,No,2,Default,20-25,Low,Small,Low Risk,Low


In [41]:
df.isnull().sum()

age                    0
income                 0
home_ownership         0
emp_length             0
loan_intent            0
loan_grade             0
loan_amount            0
interest_rate          0
loan_status            0
loan_to_income         0
prev_default           0
credit_hist_length     0
loan_status_label      0
age_group              0
income_group           0
loan_group             0
loan_risk              0
interest_rate_group    0
dtype: int64

In [39]:
df = df.dropna(subset=['age_group'])

In [43]:
df.to_csv('credit_risk_cleaned.csv', index=False)